### ISSUE TITLE
LoadUnmanagedDllFromPath / LoadUnmanagedDll is not working on linux if .so file is present in any other path other than application output directory

### ISSUE THREAD

**Issue by PrashanthIBM** [dotnet/coreclr#12707]
Hi , 

I have dependency on native(unmanaged dlls/ .so ) and currently i was setting the environment variable PATH/LD_LIBRARY_PATH variable to point to the location of unmanaged dlls.

After going through this link: https://github.com/dotnet/corefx/issues/3054

I implemented the loading of unmanaged dlls using the api "LoadUnmanagedDllFromPath" and its working on windows, 

However the same code is not working on linux even though its returning a non-zero value for library handle but still fails to load the api from unmanaged dll.

The C file(a.c) is having a simple function which returns a string. The a.c file is built into shared library a.so.

If i copy the a.so to app_output_dir/lib(/root/testClibld/bin/Debug/netcoreapp1.1/lib/a.so) and load it from this location, i get the error saying library a.so not found.

Interestingly if i copy the a.so sidebyside with application exe(/root/testClibld/bin/Debug/netcoreapp1.1/a.so), api from unmanaged dll is found.

**Environment Details:** 
Machine : Ubuntu : 14.04.1 x86_64 GNU/Linux
.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  14.04
 OS Platform: Linux
 RID:         ubuntu.14.04-x64
 Base Path:   /usr/share/dotnet/sdk/1.0.4
 
 
**a.c file contents:**
#include <stdio.h>

void getPrint()
{
    printf("Printing from the C LIbrary getPrint function");
return;
}

Attaching the a.so to quickly repro the issue.
[a.zip](https://github.com/dotnet/coreclr/files/1134647/a.zip)

**Program.cs file :** 
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Security;

using System.Reflection;
using System.Runtime.Loader;

namespace testClibld
{

       sealed internal class LoadAsm :AssemblyLoadContext
       {

          static internal String  szLoadedPath = null;
          static internal String  szLoadedLib = null;

            protected override Assembly Load(AssemblyName assemblyName)
            {

                return null;
            }

            protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
            {

               
                //szLoadedPath = "/root/testClibld/bin/Debug/netcoreapp1.1/";
                szLoadedPath = "/root/testClibld/bin/Debug/netcoreapp1.1/lib/";
                
                szLoadedLib = Path.Combine(szLoadedPath, unmanagedDllName);
            
                if(File.Exists(szLoadedLib))
                {

                    Console.WriteLine("Loading the Native dll from {0}\n ",szLoadedLib);

                }
                return LoadUnmanagedDllFromPath(szLoadedLib);
            }

            internal void Init()
            {

                IntPtr hClidrvr = LoadUnmanagedDll(@"a.so");

                if (hClidrvr != IntPtr.Zero)
                    Console.WriteLine("LIbrary handle written value is a non zero {0}\n", hClidrvr);

            }

        }
    class Program
    {

       [DllImport(@"a.so", CallingConvention = CallingConvention.StdCall, EntryPoint = "getPrint")]
       static internal extern void getPrint();
        static void Main(string[] args)
        {

            LoadAsm asm = new LoadAsm();
            asm.Init();
            Console.WriteLine("Hello World!");
            getPrint();
        }
    }

}

```
**testClibld.csproj contents:**
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
  </PropertyGroup>
<ItemGroup>

<PackageReference Include="System.Runtime.Loader" Version="4.3.0" />
</ItemGroup>

</Project>
```

**output of the program run :** 
Loading the Native dll from /root/testClibld/bin/Debug/netcoreapp1.1/lib/a.so

Library handle written value is a non zero 23358896

Hello World!

Unhandled Exception: System.DllNotFoundException: Unable to load DLL 'a.so': The specified module could not be found.
 (Exception from HRESULT: 0x8007007E)
   at testClibld.Program.getPrint()
   at testClibld.Program.Main(String[] args) in /root/prash/testClibld/Program.cs:line 82
 

Can you please let me know whether this api will work on linux or not ?

Thanks and Regards,
Prashanth.

**Comment by PrashanthIBM** [comment]
@kouvel @joshfree @pgavlin @ellismg @terrajobst @rajansingh10 @karelz  Please provide your valuable inputs.

**Comment by PrashanthIBM** [comment]
Any updates on this issue.

I have given all the required info to recreate the issue. 

Do you need anything else ?

**Comment by gkhanna79** [comment]
LoadUnmanagedDllFromPath is a simple wrapper around https://github.com/dotnet/coreclr/blob/13e7c4368da664a8b50228b1a5ef01a660fbb2dd/src/vm/assemblynative.cpp#L319 calling into Interop layer. @yizhang82 can help figure out what is going on.

**Comment by PrashanthIBM** [comment]
Hi @gkhanna79 , @yizhang82 : did you try the repro i have shared here. Any info on why it is not working ?

Is it a defect ?

**Comment by yizhang82** [comment]
I believe this is a behavior of dlopen - even though you have already loaded a.so using absolute path, the next time when you dlopen ("./a.so") it still won't find it (assuming a.so cannot be resolved using default search path rules). In other words, preloading doesn't work in linux. It only works on windows.

**Comment by PrashanthIBM** [comment]
Then how can we use this API for loading of .so ?

**Comment by yizhang82** [comment]
> Then how can we use this API for loading of .so 

It did load .so correctly. The problem is that you are using it to preload libraries so that your pinvoke would "just work". That behavior is OS specific, which is a different problem. I think the API is missing a GetProcAddress equivalent. @terrajobst 

> how it is working if we keep the .so file side by side to application exe ?

I think in order to load the native dependency correctly you need to create a nuget package for it. 
@piotrpMSFT Can you point @PrashanthIBM to existing doc/guidance on how to do that?

**Comment by jkotas** [comment]
> I think the API is missing a GetProcAddress equivalent.

Why do you think so? I do not think this issue would be helped by having GetProcAddress equivalent here.

> to load the native dependency correctly you need to create a nuget package for it

Right. The native dependencies needs to follow the conventions to make them "just work".

**Comment by yizhang82** [comment]
> Why do you think so? I do not think this issue would be helped by having GetProcAddress equivalent here.

Sorry. I didn't explain myself properly. If there is a GetProcAddress equivalent, he can simply call that on the handle returned from LoadUnmanagedDllFromPath and convert that into a delegate which he can now call as a pinvoke delegate. 

Probably not necessary in this case if he follows the convention to create nuget package for native dependency, but I can imagine there will be cases that having a x-plat LoadLibrary/GetprocAddress equivalent would be helpful for custom loading logic. We only have half of the story today.

**Comment by jkotas** [comment]
> he can simply call that on the handle returned from LoadUnmanagedDllFromPath

`LoadUnmanagedDllFromPath` is protected - you cannot simply call it from outside. This API was not meant to solve the scenario discussed here.

**Comment by PrashanthIBM** [comment]
ok can you help me how to solve my scenario ?. 

It is working fine with windows . Even on Linux loading of .so is happening but after that if i use any api calls present in the .so, it is not be able to find.

Dont you think default search path for the .so must be one of the loaded path(absolute path returned from LoadUnmanagedDllFromPath) and must be first in the default search path list ?

What is the use of this api if it will not enable me to call the functions at a later stage in application ?

**Comment by PrashanthIBM** [comment]
ok can you help me how to solve my scenario ?. 

It is working fine with windows . Even on Linux loading of .so is happening but after that if i use any api calls present in the .so, it is not be able to find.

Dont you think default search path for the .so must be one of the loaded path(absolute path returned from LoadUnmanagedDllFromPath) and must be first in the default search path list ?

What is the use of this api if it will not enable me to call the functions at a later stage in application ?

**Comment by jkotas** [comment]
> What is the use of this api if it will not enable me to call the functions at a later stage in application ?

https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md is documentation for AssemblyLoadContext that this API is part of. This API is designed to supply native dependencies of isolated plugins loaded in its own AssemblyLoadContext. It is not designed to allow supplying native libraries for the default load context.

> ok can you help me how to solve my scenario ?

Why is it not an option for you to just drop the native library into your application directory? Why it has to be in the `lib` subdirectory?

If you have a good reason for loading native libraries from arbitrary paths, it has to be done today by PInvoking dlopen and dlsym; and wrapping the pointer using `Marshal.GetDelegateForFunctionPointer`.

**Comment by PrashanthIBM** [comment]
@jkotas  I will not be able to copy the libraries to application bin path, since we have lot of library dependencies and other folder structure dependencies. 

Is there no way to append the default search path with the loaded path ?

**Comment by jkotas** [comment]
> Is there no way to append the default search path with the loaded path ?

You can build your own version of the host executable that has your custom .so probing paths set. e.g. using RPATH.

The source for the host executable are at https://github.com/dotnet/core-setup/tree/master/src/corehost.

**Comment by PrashanthIBM** [comment]
ok i will look into it. 

But how is it working on windows ? Can i consider using this solution  on windows at least ? 

Also dont you think there should be something similar to Full .Net Framework LoadLibrary in .Net Core which will load the native dlls for the life time of application process instead of with in the loadcontext and the loaded dlls are available till the end of the application ?

**Comment by hovishwa** [comment]
Going by this discussion, it looks like nobody uses LoadLibrary/PInvoke on Linux to load and call any functions without a package. Is MS suggesting this is the way(using a package) to go in future?  There may be lot of components in the market which uses LoadLibrary and PInvoke, MS has any guidelines for these scenarios?   If the size of such native package is more ( >200MB), do you have any suggestions? Thank you.

**Comment by yizhang82** [comment]
@PrashanthIBM 

> But how is it working on windows ? 

As I've explained earlier, If you try to load a DLL without path for a 2nd time, Windows will simply return the already loaded DLL without searching. Please refer to https://msdn.microsoft.com/en-us/library/windows/desktop/ms682586(v=vs.85).aspx, in particular:

```
Before the system searches for a DLL, it checks the following:
If a DLL with the same module name is already loaded in memory, the system uses the loaded DLL, no matter which directory it is in. The system does not search for the DLL.
```

This behavior is specific to Windows. Linux doesn't do this and will always search for the .so again, at least according to my late night experiment on Ubuntu. 

> Can i consider using this solution on windows at least ?

You are free to choose whatever solution works best for you. :) But if you want your code to work x-plat, you need to either create a nuget package or create a custom host with custom probing path as @jkotas suggested earlier. 

> Also dont you think there should be something similar to Full .Net Framework LoadLibrary in .Net Core which will load the native dlls for the life time of application process instead of with in the loadcontext and the loaded dlls are available till the end of the application ?

The DLL/so is indeed loaded for the lifetime of the application, it is the behavior of OS loadlibrary/dlopen implementation that affects whether the 2nd loadlibrary/dlopen (done by the p/invoke) succeeds or not. More specifically, .NET is relying on OS to make the call that /your/absolutely/path/a.so and a.so is the same library. One could argue that .NET should've decided that a.so is /your/absolutely/path/a.so and simply use that, but I believe this is a policy that is best left to the OS to decide.

**Comment by PrashanthIBM** [comment]
@jkotas  @yizhang82 :  

So seems like i can use this solution only on windows.
 
Let me explain my scenario: 

1) We have a database application which will connect to database through .Net Core driver.

2) .Net Core driver inturn depends on native C driver to connect which is not a single library(.so file) and it is having multiple .so file and dependencies on other contents files(like msg files, config files, trace exe ..) 



@jkotas  i tried appending the LD_LIBRARY_PATH system env variable using the System.Environment at the process level and even this did not work.

The following code is added in .Net Core driver :
szLoadedPath = "/root/testClibld/bin/Debug/netcoreapp1.1/lib/";
Environment.SetEnvironmentVariable("LD_LIBRARY_PATH", szLoadedDB2Path);

Any suggestion on how to make this work ?

**Comment by jkotas** [comment]
> appending the `LD_LIBRARY_PATH` system env variable using the System.Environment at the process level and even this did not work

`LD_LIBRARY_PATH` has to be set when the process starts. Setting it half way through won't have the desired effect. This is how Unix .so loader works. Nothing to do with .NET.

A [common trick](https://stackoverflow.com/questions/13685042/how-to-set-a-unix-dynamic-library-path-ld-library-path-in-java) around this problem is to re-execute yourself: set the `LD_LIBRARY_PATH` environment variable to what you need - if it is not set correctly; and then start a new copy of your app as a child process. It is very similar to just setting the env. variable in a wrapper shell-script, except that you avoid having multiple executables.

**Comment by PrashanthIBM** [comment]
Hi @jkotas @yizhang82 

Cant we do static loading of dlls in linux and keep them alive until the parent process which loaded the dll is alive ?

The scenario which we are facing is that Application is agnostic of the unmanaged dlls and the application can talk to only the managed database provider and internally in the managed database provider, the dependent unmanaged dll is being loaded using the custom load context. 

Do we you any taught's on this  approach ?

**Comment by kouvel** [comment]
I'm not too familiar with this area but when the application has a managed dependency that in turn has native dependencies, my understanding is that the options are as stated above: put the managed/native dependencies alongside the app, create a package for the managed/native dependencies, create a custom host, or p/invoke. LoadUnmanagedDllFromPath may be used with the option of a custom host to provide custom loading logic at run time for unmanaged dependencies, or custom loading paths could be set by the host instead along with using DllImport, but with my limited understanding I don't think they can be combined.

**Comment by kouvel** [comment]
> I don't think they can be combined

To clarify what I believe, LoadUnmanagedDllFromPath would come into play during DllImport's loading logic when the default AssemblyLoadContext is changed, which requires a custom host. Alternatively, the default load paths can be changed by a custom host to have DllImport find the native dependencies. Alternatively, the native dependencies could be located in the default host's search locations (alongside the app for instance), and that would work with the default host.

**Comment by jkotas** [comment]
We should consider adding a callback that allow folks to supply custom resolution policy for unmanaged libraries. We have intended to do that with `AssemblyLoadContext.LoadUnmanagedDll`. At the end, it only addressed a small fraction of them problem because of it does not work for default load context. When we have introduced `AssemblyLoadContext.LoadUnmanagedDll` originally, we allowed overriding the default load context as well but then removed this capability and did not think through the impact on all scenarios.

cc @russellhadley
